File: src/content/docs/topics/cs/coding-problems/2d-dynamic-programming/309-best-time-to-buy-and-sell-stock-with-cooldown-approach3.go
package main

import "fmt"

func maxProfit(prices []int) int {
	max := func(a, b int) int {
		if a > b {
			return a
		}
		return b
	}
	if len(prices) == 0 {
		return 0
	}
	hold := -prices[0]
	sold := 0
	rest := 0
	for i := 1; i < len(prices); i++ {
		prevHold, prevSold, prevRest := hold, sold, rest
		hold = max(prevHold, prevRest-prices[i])
		sold = prevHold + prices[i]
		rest = max(prevRest, prevSold)
	}
	return max(sold, rest)
}

func main() {
	assert := func(cond bool) {
		if !cond {
			panic("assertion failed")
		}
	}
	assert(maxProfit([]int{1, 2, 3, 0, 2}) == 3)
	assert(maxProfit([]int{1}) == 0)
	assert(maxProfit([]int{}) == 0)
	assert(maxProfit([]int{5, 4, 3, 2, 1}) == 0)
	assert(maxProfit([]int{1, 2, 3, 4, 5}) == 4)
	assert(maxProfit([]int{1, 2}) == 1)
	fmt.Println("all tests pass")
}

File: src/content/docs/topics/cs/coding-problems/2d-dynamic-programming/309-best-time-to-buy-and-sell-stock-with-cooldown-approach3.py
def max_profit(prices):
    if not prices:
        return 0
    hold = -prices[0]
    sold = 0
    rest = 0
    for i in range(1, len(prices)):
        prev_hold, prev_sold, prev_rest = hold, sold, rest
        hold = max(prev_hold, prev_rest - prices[i])
        sold = prev_hold + prices[i]
        rest = max(prev_rest, prev_sold)
    return max(sold, rest)

assert max_profit([1, 2, 3, 0, 2]) == 3
assert max_profit([1]) == 0
assert max_profit([]) == 0
assert max_profit([5, 4, 3, 2, 1]) == 0
assert max_profit([1, 2, 3, 4, 5]) == 4
assert max_profit([1, 2]) == 1
print("all tests pass")
